[triton][sigmoid] Fix kernel cache and serialization issue for triton sigmoid + CUDA kernel bug

shawnxu26 opened the pull request:
Differential Revision: D84930778

minjang commented on torch/_export/serde/serialize.py:
```diff
@@ -552,17 +539,103 @@ def get_triton_kernel_and_cache_entry(node: torch.fx.Node):
+    if len(matching_entries) == 1:
+        return kernel, matching_entries[0][1]
+
+    if len(matching_entries) > 1:
```
nit. just coding style suggestion. you may early raise if the entries is empty. then you can reduce indentations for a large chunk of the code.

minjang reviewed (approved):
LGTM. A minor coding style suggestion and number format in constexpr.

XueningXu commented on torch/_export/serde/serialize.py:
```diff
@@ -552,17 +539,108 @@ def get_triton_kernel_and_cache_entry(node: torch.fx.Node):
+        matching_entries = []
+        for sig_key, cache_entry in cache.items():
+            constexpr_matches = re.findall(
+                r"\('constexpr',\s*(-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?)\)", sig_key
```
make sense, changed to `[+-]?`

XueningXu commented on torch/_export/serde/serialize.py:
```diff
@@ -552,17 +539,108 @@ def get_triton_kernel_and_cache_entry(node: torch.fx.Node):
+                if hasattr(config, "num_ctas") and hasattr(entry_metadata, "num_ctas"):
+                    if config.num_ctas != entry_metadata.num_ctas:
+                        metadata_matches = False
+                        continue
```
Thanks for the comments! I've refactored the code as suggested.

shawnxu26 committed:
[triton][sigmoid] Fix kernel cache and serialization issue for triton sigmoid + CUDA kernel bug (#166098)

Summary:

## For fetching correct kernel cache
- Step1: constexpr info is included in graph nodes, and their combinations can be used to uniquely identify each kernel cache, which also makes more sense in terms of why we are having multiple kernel caches.
- Step2: In case two kernel caches have same (or empty) constexpr but different autotuner configs, need further matching on autotuner configs to find the correct kernel cache.
- The signature formats for CPU and GPU are the same on the new `triton/experimental`:
    -  GPU and CPU: `"[('*fp32', 'D'), ('*fp32', 'D'), ('*fp32', 'D'), ('i32', ''), ('constexpr', 16)]{'debug': False}"`


TODO: The Triton compiler specializes kernel arguments when generating kernel cache entry signatures, especially for integers (treating integer 1 as constexpr). This specialization behavior necessitates a more sophisticated matching strategy to correctly identify kernel cache entries for a given graph node in the future. ([jit.py](https://fburl.com/code/3n7sgd3u), [compiler.py](https://fburl.com/code/n6v5h5k8)) Currently, for some transform kernels (e.g., ngram), we wrap scalar arguments as tensors to avoid integer being treated as constexpr unexpectedly.

## Tried to serialize file twice
- This is essentially due to the naming of the kernel cache. Originally, all kernel caches are named based on `kernel_name` (e.g., `add_kernel.so`), without adding any name mangling.
- The fix was to append the unique hash to the kernel name: `add_kernel_<unique_hash>.so`.
- Add `symbol_name` for `dlsym()` to find correct kernel name (without hash).

## Segmentation fault for launching CUDA kernel `cuLaunchKernel`
It was crashing with a segmentation fault (SIGSEGV at address `0x0`) when attempting to launch Triton-compiled CUDA kernels via `cuLaunchKernel`. The crash occurred inside the CUDA driver when it tried to read kernel parameters from the `args` array.
- Taking [`add_kernel`](https://fburl.com/code/19qw6qyp) as an example, the kernel expects 4 parameters excluding `constexpr`.
- An additional `global_scratch` parameter is baked into kernel during compilation, especially TritonGPU --> LLVM IR ([See code here](https://fburl.com/code/h97maspq)). This can also be observed in [`driver.py`](https://fburl.com/code/ufotoekq), Triton-compiled kernels automatically append an additional parameter `&global_scratch`.
- The bug in the previous implementation is that when [preparing kernel inputs](https://fburl.com/code/2tnlx9nm), it does not include this additional parameter.
- As a result, when `cuLaunchKernel` executes, the CUDA driver tries to read the fifth argument `args[4]` to get `global_scratch` parameter, but this accesses garbage memory that causes a NULL pointer dereference and segmentation fault.

Test Plan:
Now testing on `triton/experimental`

## CUDA
Passed
```
buck run @//mode/opt -c remoteexecution.local=enabled fbcode//sigmoid/inference/test:test_passes -- -r test_triton_hop_cuda
```
---
## CPU

Passed
```
buck run @//mode/opt -c remoteexecution.local=enabled fbcode//sigmoid/inference/test:test_passes -- -r test_triton_hop_cpu
```
WARNING: missing `.launcher.so` file for cpu backend on `triton/experimental`.

Reviewed By: minjang

Differential Revision: D84930778

shawnxu26 committed:
Merge branch 'pytorch:main' into export-D84930778

shawnxu26 committed:
Merge branch 'pytorch:main' into export-D84930778
